napari/components/_dims/dims.py

```diff
@@ -0,0 +1,397 @@
+    range : list
+        List of tuples (min, max, step), one for each dimension
+    point : list
+        List of floats, one for each dimension
+    interval : list
+        List of tuples (min, max), one for each dimension
+    mode : list
+        List of DimsMode, one for each dimension
+    display : list
+        List of bool indicating if dimension displayed or not, one for each
+        dimension
+    ndims : int
+        Number of dimensions
+    displayed : np.ndarray
+        Array of the displayed dimensions
+    """
+    def __init__(self, init_ndims=0):
+        super().__init__()
+
+        # Events:
+        self.events = EmitterGroup(source=self,
+                                    auto_connect=True,
+                                    axis=None,
+                                    ndims=None)
+
+        self.range = []
+        self.point = []
+        self.interval = []
+        self.mode = []
+        self.display = []
+
+        self._ensure_axis_present(init_ndims - 1)
+
+    def __str__(self):
+        return "~~".join([str(self.range),
+                         str(self.point),
+                         str(self.interval),
+                         str(self.mode),
+                         str(self.display)])
+
+    @property
+    def ndims(self):
+        """Returns the number of dimensions
+
+        Returns
+        -------
+        ndims : int
+            Number of dimensions
+        """
+        return len(self.point)
+
+    @ndims.setter
+    def ndims(self, ndims):
+        if self.ndims < ndims:
+            self._ensure_axis_present(ndims - 1)
+        elif self.ndims > ndims:
+            self._trim_ndims(ndims)
+
+    @property
+    def displayed(self):
+        """Returns the displayed dimensions
+
+        Returns
+        -------
+        displayed : np.ndarray
+            Displayed dimensions
+        """
+        displayed_one_hot = copy(self.display)
+        displayed_one_hot =  ([False if elem is None else elem for elem in
+                              displayed_one_hot])
+        return np.nonzero(list(displayed_one_hot))[0]
+
+    @property
+    def slice_and_project(self):
+        """Returns the slice and project tuples that specify how to slice and
+        project arrays.
+
+        Returns
+        -------
+        slice : tuple
+            The slice tuple
+        project : tuple
+            The projection tuple
+        """
+
+        slice_list = []
+        project_list = []
+
+        for (mode, display, point, interval, range) in zip(self.mode, self.display, self.point, self.interval, self.range):
+
+            if mode == DimsMode.Point or mode is None:
+                if display:
+                    # no slicing, cropping or projection:
+                    project_list.append(False)
+                    slice_point = round(point)
+                    slice_point = 1 if slice_point==0 else slice_point
+                    ## TODO: workaround here...
```

Comment:
I am working to make this mergeable - and unsurprisingly have narrowed down to this spot as the location for the current bug I am trying to fix ....